File: streampark-common/src/main/scala/org/apache/streampark/common/util/HdfsUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.streampark.common.util

import org.apache.commons.codec.digest.DigestUtils
import org.apache.hadoop.fs._
import org.apache.hadoop.hdfs.DistributedFileSystem
import org.apache.hadoop.io.IOUtils
import org.apache.hadoop.ipc.RPC

import java.io.{ByteArrayOutputStream, FileWriter, IOException}
import java.net.InetSocketAddress

import scala.util.{Failure, Success, Try}

object HdfsUtils extends Logger {

  def getDefaultFS: String =
    HadoopUtils.hadoopConf.get(FileSystem.FS_DEFAULT_NAME_KEY)

  def list(src: String): List[FileStatus] =
    HadoopUtils.hdfs.listStatus(getPath(src)).toList

  def move(src: String, dst: String): Unit =
    HadoopUtils.hdfs.rename(getPath(src), getPath(dst))

  def mkdirs(path: String): Unit = HadoopUtils.hdfs.mkdirs(getPath(path))

  def copyHdfs(
      src: String,
      dst: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit = {
    val srcPath = getPath(src)
    val dstPath = getPath(dst)
    val dstStatus = HadoopUtils.hdfs.getFileStatus(dstPath)
    val dstFinalPath =
      if (dstStatus.isFile) dstPath
      else {
        getPath(s"$dst/${srcPath.getName}")
      }
    FileUtil.copy(
      HadoopUtils.hdfs,
      srcPath,
      HadoopUtils.hdfs,
      dstFinalPath,
      delSrc,
      overwrite,
      HadoopUtils.hadoopConf)
  }

  def copyHdfsDir(
      src: String,
      dst: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit = {
    list(src).foreach(x =>
      FileUtil.copy(
        HadoopUtils.hdfs,
        x,
        HadoopUtils.hdfs,
        getPath(dst),
        delSrc,
        overwrite,
        HadoopUtils.hadoopConf))
  }

  def upload(src: String, dst: String, delSrc: Boolean = false, overwrite: Boolean = true): Unit =
    HadoopUtils.hdfs.copyFromLocalFile(delSrc, overwrite, getPath(src), getPath(dst))

  def uploadMulti(
      src: Array[String],
      dst: String,
      delSrc: Boolean = false,
      overwrite: Boolean = true): Unit =
    HadoopUtils.hdfs.copyFromLocalFile(delSrc, overwrite, src.map(getPath), getPath(dst))

  def download(
      src: String,
      dst: String,
      delSrc: Boolean = false,
      useRawLocalFileSystem: Boolean = false): Unit =
    HadoopUtils.hdfs.copyToLocalFile(delSrc, getPath(src), getPath(dst), useRawLocalFileSystem)

  def getNameNode: String = Try(getAddressOfActive(HadoopUtils.hdfs).getHostString) match {
    case Success(value) => value
    case Failure(exception) => throw exception
  }

  /**
   * create a new file on hdfs, write some data to it
   *
   * @param fileName
   * @param content
   *   \@throws
   */
  def create(fileName: String, content: String): Unit = {
    val path: Path = getPath(fileName)
    require(HadoopUtils.hdfs.exists(path), s"[StreamPark] HdfsUtils.create $fileName is exists!! ")
    val outputStream: FSDataOutputStream = HadoopUtils.hdfs.create(path)
    outputStream.writeUTF(content)
    outputStream.flush()
    outputStream.close()
  }

  def exists(path: String): Boolean = HadoopUtils.hdfs.exists(getPath(path))

  def read(fileName: String): String = {
    val path: Path = getPath(fileName)
    require(
      HadoopUtils.hdfs.exists(path) && !HadoopUtils.hdfs.isDirectory(path),
      s"[StreamPark] HdfsUtils.read: path($fileName) not exists or isDirectory ")
    val in = HadoopUtils.hdfs.open(path)
    val out = new ByteArrayOutputStream()
    IOUtils.copyBytes(in, out, 4096, false)
    out.flush()
    IOUtils.closeStream(in)
    IOUtils.closeStream(out)
    new String(out.toByteArray)
  }

  def delete(src: String): Unit = {
    val path: Path = getPath(src)
    if (HadoopUtils.hdfs.exists(path)) {
      HadoopUtils.hdfs.delete(path, true)
    } else {
      logWarn(s"HDFS delete $src, but file $src is not exists!")
    }
  }

  def fileMd5(fileName: String): String = {
    val path = getPath(fileName)
    val in = HadoopUtils.hdfs.open(path)
    Try(DigestUtils.md5Hex(in)) match {
      case Success(s) =>
        in.close()
        s
      case Failure(e) =>
        in.close()
        throw e
    }
  }

  def downToLocal(hdfsPath: String, localPath: String): Unit = {
    val path: Path = getPath(hdfsPath)
    val input: FSDataInputStream = HadoopUtils.hdfs.open(path)
    val content: String = input.readUTF
    val fw: FileWriter = new FileWriter(localPath)
    fw.write(content)
    fw.close()
    input.close()
  }

  private[this] def getPath(hdfsPath: String) = new Path(hdfsPath)

  @throws[IOException]
  def getAddressOfActive(fs: FileSystem): InetSocketAddress = {
    if (!fs.isInstanceOf[DistributedFileSystem]) {
      throw new IllegalArgumentException(s"FileSystem $fs is not a DFS.")
    }
    // force client address resolution.
    fs.exists(new Path("/"))
    val dfs = fs.asInstanceOf[DistributedFileSystem]
    val dfsClient = dfs.getClient
    RPC.getServerAddress(dfsClient.getNamenode)
  }

}

File: streampark-common/src/main/java/org/apache/streampark/common/util/HdfsUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.util;

import org.apache.streampark.shaded.org.slf4j.Logger;

import org.apache.commons.codec.digest.DigestUtils;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FSDataOutputStream;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.FileUtil;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hdfs.DistributedFileSystem;
import org.apache.hadoop.io.IOUtils;
import org.apache.hadoop.ipc.RPC;

import java.io.ByteArrayOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.Arrays;
import java.util.List;

public final class HdfsUtils {

    private static final Logger LOG =
        StreamParkLoggerFactory.loggerFactory().getLogger(HdfsUtils.class.getName());

    private HdfsUtils() {
    }

    public static String getDefaultFS() {
        return HadoopUtils.hadoopConf().get(FileSystem.FS_DEFAULT_NAME_KEY);
    }

    public static List<FileStatus> list(String src) throws IOException {
        return Arrays.asList(HadoopUtils.hdfs().listStatus(getPath(src)));
    }

    public static void move(String src, String dst) throws IOException {
        HadoopUtils.hdfs().rename(getPath(src), getPath(dst));
    }

    public static void mkdirs(String path) throws IOException {
        HadoopUtils.hdfs().mkdirs(getPath(path));
    }

    public static void copyHdfs(String src, String dst) throws IOException {
        copyHdfs(src, dst, false, true);
    }

    public static void copyHdfs(String src, String dst, boolean delSrc, boolean overwrite) throws IOException {
        Path srcPath = getPath(src);
        Path dstPath = getPath(dst);
        FileStatus dstStatus = HadoopUtils.hdfs().getFileStatus(dstPath);
        Path dstFinalPath =
            dstStatus.isFile() ? dstPath : getPath(dst + "/" + srcPath.getName());
        FileUtil.copy(
            HadoopUtils.hdfs(),
            srcPath,
            HadoopUtils.hdfs(),
            dstFinalPath,
            delSrc,
            overwrite,
            HadoopUtils.hadoopConf());
    }

    public static void copyHdfsDir(String src, String dst) throws IOException {
        copyHdfsDir(src, dst, false, true);
    }

    public static void copyHdfsDir(String src, String dst, boolean delSrc, boolean overwrite) throws IOException {
        for (FileStatus status : list(src)) {
            FileUtil.copy(
                HadoopUtils.hdfs(),
                status,
                HadoopUtils.hdfs(),
                getPath(dst),
                delSrc,
                overwrite,
                HadoopUtils.hadoopConf());
        }
    }

    public static void upload(String src, String dst) throws IOException {
        upload(src, dst, false, true);
    }

    public static void upload(String src, String dst, boolean delSrc, boolean overwrite) throws IOException {
        HadoopUtils.hdfs().copyFromLocalFile(delSrc, overwrite, getPath(src), getPath(dst));
    }

    public static void uploadMulti(String[] src, String dst) throws IOException {
        uploadMulti(src, dst, false, true);
    }

    public static void uploadMulti(String[] src, String dst, boolean delSrc, boolean overwrite) throws IOException {
        Path[] paths = Arrays.stream(src).map(HdfsUtils::getPath).toArray(Path[]::new);
        HadoopUtils.hdfs().copyFromLocalFile(delSrc, overwrite, paths, getPath(dst));
    }

    public static void download(String src, String dst) throws IOException {
        download(src, dst, false, false);
    }

    public static void download(String src, String dst, boolean delSrc,
                                boolean useRawLocalFileSystem) throws IOException {
        HadoopUtils.hdfs()
            .copyToLocalFile(delSrc, getPath(src), getPath(dst), useRawLocalFileSystem);
    }

    public static String getNameNode() throws IOException {
        return getAddressOfActive(HadoopUtils.hdfs()).getHostString();
    }

    public static void create(String fileName, String content) throws IOException {
        Path path = getPath(fileName);
        if (!HadoopUtils.hdfs().exists(path)) {
            throw new IllegalArgumentException("[StreamPark] HdfsUtils.create " + fileName + " is exists!! ");
        }
        FSDataOutputStream outputStream = HadoopUtils.hdfs().create(path);
        outputStream.writeUTF(content);
        outputStream.flush();
        outputStream.close();
    }

    public static boolean exists(String path) throws IOException {
        return HadoopUtils.hdfs().exists(getPath(path));
    }

    public static String read(String fileName) throws IOException {
        Path path = getPath(fileName);
        if (!HadoopUtils.hdfs().exists(path) || HadoopUtils.hdfs().isDirectory(path)) {
            throw new IllegalArgumentException(
                "[StreamPark] HdfsUtils.read: path(" + fileName + ") not exists or isDirectory ");
        }
        FSDataInputStream in = HadoopUtils.hdfs().open(path);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        IOUtils.copyBytes(in, out, 4096, false);
        out.flush();
        IOUtils.closeStream(in);
        IOUtils.closeStream(out);
        return new String(out.toByteArray());
    }

    public static void delete(String src) throws IOException {
        Path path = getPath(src);
        if (HadoopUtils.hdfs().exists(path)) {
            HadoopUtils.hdfs().delete(path, true);
        } else {
            LOG.warn("[StreamPark] HDFS delete {}, but file {} is not exists!", src, src);
        }
    }

    public static String fileMd5(String fileName) throws IOException {
        Path path = getPath(fileName);
        FSDataInputStream in = HadoopUtils.hdfs().open(path);
        try {
            // MD5 is used for non-cryptographic file integrity checks only.
            @SuppressWarnings("java:S4790")
            String digest = DigestUtils.md5Hex(in);
            return digest;
        } finally {
            in.close();
        }
    }

    public static void downToLocal(String hdfsPath, String localPath) throws IOException {
        Path path = getPath(hdfsPath);
        FSDataInputStream input = HadoopUtils.hdfs().open(path);
        String content = input.readUTF();
        FileWriter fw = new FileWriter(localPath);
        fw.write(content);
        fw.close();
        input.close();
    }

    private static Path getPath(String hdfsPath) {
        return new Path(hdfsPath);
    }

    public static InetSocketAddress getAddressOfActive(FileSystem fs) throws IOException {
        if (!(fs instanceof DistributedFileSystem)) {
            throw new IllegalArgumentException("FileSystem " + fs + " is not a DFS.");
        }
        fs.exists(new Path("/"));
        DistributedFileSystem dfs = (DistributedFileSystem) fs;
        return RPC.getServerAddress(dfs.getClient().getNamenode());
    }
}
